File: java/0704.0_Binary_Search.java
/*
T: O(logN)
S: O(1)

Runtime: 0 ms, faster than 100.00% of Java online submissions for Binary Search.
Memory Usage: 54.3 MB, less than 51.69% of Java online submissions for Binary Search.
*/

class Solution {
    public int search(int[] nums, int target) {
        int pivot, left = 0, right = nums.length - 1;
        while (left <= right) {
            pivot = left + (right - left) / 2;
            if (nums[pivot] == target) return pivot;
            if (target < nums[pivot]) right = pivot - 1;
            else left = pivot + 1;
        }   
        return -1;
    }
}

File: js/0704.0_Binary_Search.js
/*
T: O(logN)
S: O(1)

Runtime: 78 ms, faster than 68.45% of JavaScript online submissions for Binary Search.
Memory Usage: 45.7 MB, less than 7.01% of JavaScript online submissions for Binary Search.
*/

/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var search = function(nums, target) {
    let left = 0;
    let right = nums.length - 1;
    while (left <= right) {
        const middle = Math.floor((left + right) / 2);
        if (nums[middle] > target) {
            right = middle - 1;
        } else if (nums[middle] < target) {
            left = middle + 1;
        } else {
            return middle;
        }
    }
    return -1;
};
